Instantiating grains that implement only generic interfaces

ca-ta opened the pull request:
Fixes #338 

All tests in TestAll.cmd pass (and my grains now work, yay :-) ).

Notes:
* One can now instantiate *non-generic* grains that directly implement one or more generic interfaces
* For *generic* grain classes, the constraints that were already implied in the code still apply (i.e. the generic class must take the same generic arguments as the implemented generic interface)
 * note that I am unsure if that particular code path was ever hit - I think that scenario was also broken (I have not tested it, though).

Please review, especially the change to ```GrainClassData```. I have noticed that it is serializable and I have marked the newly introduced field as ```[NonSerialized]``` in an attempt to not break a rolling update scenario. I've followed what seems to be an existing pattern, but I am not in love with that particular bit - while it works (the relevant method is only called from the ```Catalog```), it feels quite brittle.

gabikliot commented:
@ca-ta , can you please also add the grain and the test that uses it, to the Tester project, so we can have it in our tests. It will also help "formally" document the supported scenario.

sergeybykov commented on src/Orleans/Runtime/GrainInterfaceMap.cs:
```diff
@@ -217,6 +223,15 @@ public bool TryGetGrainTypeCode(int grainInterfaceId, out int grainTypeCode, str
             return TryGetGrainTypeCode(interfaceData, out grainTypeCode, grainClassNamePrefix);
         }
 
+        private string typeKey(Type interfaceType)
```
Should be TypeKey (Pascal style) or even better if GetTypeKey.

ca-ta commented:
@gabikliot : Will do.
@sergeybykov: Indeed - sorry about that, I've been writing a lot of F# lately.

Thanks guys, I'll update the PR shortly.

sergeybykov commented:
@ca-ta Thanks! Looks good. Doesn't seem to break any of our internal tests.

I second @gabikliot - an example in a form of a test would help document the intended usage and protect from breaking in the future.

As I understand it can be as simple as:
```csharp
public interface IGenericGrainForNonGenericClass<T> : IGrainWithIntegerKey
{
   Task<T> Ping(T t);
}

public class NonGenericGrainForGenericInterface : Grain, IGenericGrainForNonGenericClass<int>
{
    public Task<int> Ping(int t)
    {
         return Task.FromResult(t+1);
    }
}

var g = GrainFactory.GetGrain<IGenericGrainForNonGenericClass<int>>(0);
int x = await g.Ping(5);
Assert.AreEqual(6, x);
```

But maybe you had more in mind or can add more sophisticated test cases.

gabikliot commented:
and also:
```csharp
var g = GrainFactory.GetGrain<NonGenericGrainForGenericInterface>(1);
int x = await g.Ping(5);
Assert.AreEqual(6, x);
```

right?

gabikliot commented:
I deleted my previous comment. It was a different case. Don't want to confuse anyone.

ca-ta commented:
I've added tests, and while at it discovered that instantiating generic grain classes didn't work (I've possibly broken it in my previous commit). Anyway, I've added tests for that scenario too and now both scenarios work:
* instantiating non-generic grains that implement generic interfaces
* instantiating generic grains that:
  * implement a generic interface AND
  * take the same generic arguments as the implemented generic interface.

ca-ta commented:
Hold this PR - It needs further testing.

ca-ta commented:
I have simplified the grain implementation resolution so that, if both generic and concrete implementations exist for a given generic interface, the concrete grain implementation is preferred. The previous implementation would throw in this scenario - I found that counter-intuitive. 

I've also added a corresponding test. All available Orleans test pass as well.

To recap, the grain type resolution of grains implementing generic interfaces is:
 1. First, try to locate a concrete implementation. 
    * existing rules apply if multiple matching implementations exist.
    * if an unambiguous implementation is found, it is returned and no further type resolution is performed.

 2. If a concrete implementation is not found, try to locate a generic class implementing the interface and instantiate a specialization of that class.
    * the generic class must take the same generic arguments as the generic interface (this restriction was already in place, and is reasonable)
    * existing rules apply if multiple matching implementations exist.

Sorry for the three commits - let me know if you'd prefer to close this pull request and re-submit as one unified commit.

Also, please let me know if you think there are any gotchas in regards to the chosen grain type resolution strategy. It works great for me but it may have side-effects I am not aware of.

ca-ta commented:
PR ready to go. - I've also removed the "hold" comment

ca-ta commented:
I'll close this and add a pull request with only one commit.
